lib/archethic/db/embedded_impl/chain_reader.ex

```diff
@@ -266,43 +307,64 @@ defmodule Archethic.DB.EmbeddedImpl.ChainReader do
       ChainIndex.list_chain_addresses(genesis_address, db_path)
       |> Enum.map(&elem(&1, 0))
 
-    {limit_address, paging_state, more?, new_paging_state} =
+    {nb_to_take, paging_state, more?, new_paging_state} =
       case Keyword.get(opts, :paging_state) do
         nil ->
           chain_length = Enum.count(all_addresses_asc)
 
           if chain_length <= @page_size do
-            {nil, nil, false, ""}
+            {@page_size, nil, false, nil}
           else
             idx = chain_length - 1 - @page_size
 
-            {nil, all_addresses_asc |> Enum.at(idx), true, all_addresses_asc |> Enum.at(idx + 1)}
+            {@page_size, all_addresses_asc |> Enum.at(idx), true,
+             all_addresses_asc |> Enum.at(idx + 1)}
           end
 
         paging_state ->
           paging_state_idx =
             all_addresses_asc
             |> Enum.find_index(&(&1 == paging_state))
 
-          limit_address = Enum.at(all_addresses_asc, paging_state_idx - 1)
-
           if paging_state_idx < @page_size do
-            {limit_address, nil, false, ""}
+            {paging_state_idx, nil, false, nil}
           else
             idx = paging_state_idx - 1 - @page_size
 
-            {limit_address, all_addresses_asc |> Enum.at(idx), true,
+            {@page_size, all_addresses_asc |> Enum.at(idx), true,
              all_addresses_asc |> Enum.at(idx + 1)}
```

Comment:
for clarity, I would suggest to do the transformations before returning the tuple